From bfefd6a26d574209503e27ea50b9cea933ca370e Mon Sep 17 00:00:00 2001
From: Stefan Berger <stefanb@linux.ibm.com>
Date: Wed, 6 Apr 2022 14:16:09 -0400
Subject: [PATCH 36/90] ima: Add ima-ns template that also logs the user
 namespace UUID

Add ima-ns template that also logs the user namespace's UUID along with
measurements, file, and file signature.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>
---
 security/integrity/ima/Kconfig        | 6 ++++++
 security/integrity/ima/ima_template.c | 1 +
 2 files changed, 7 insertions(+)

diff --git a/security/integrity/ima/Kconfig b/security/integrity/ima/Kconfig
index e5db9781441f..2d2433cd7175 100644
--- a/security/integrity/ima/Kconfig
+++ b/security/integrity/ima/Kconfig
@@ -71,11 +71,16 @@ choice
 	  template permits both larger hash digests and longer
 	  pathnames. The configured default template can be replaced
 	  by specifying "ima_template=" on the boot command line.
+	  The 'ima-ns' template logs the UUID of the user namespace
+	  from where a measurement originated.
 
 	config IMA_NG_TEMPLATE
 		bool "ima-ng (default)"
 	config IMA_SIG_TEMPLATE
 		bool "ima-sig"
+	config IMA_NS_TEMPLATE
+		bool "ima-ns"
+		depends on IMA_NS
 endchoice
 
 config IMA_DEFAULT_TEMPLATE
@@ -83,6 +88,7 @@ config IMA_DEFAULT_TEMPLATE
 	depends on IMA
 	default "ima-ng" if IMA_NG_TEMPLATE
 	default "ima-sig" if IMA_SIG_TEMPLATE
+	default "ima-ns" if IMA_NS_TEMPLATE
 
 choice
 	prompt "Default integrity hash algorithm"
diff --git a/security/integrity/ima/ima_template.c b/security/integrity/ima/ima_template.c
index c950d2a2ce07..8e02ee151d87 100644
--- a/security/integrity/ima/ima_template.c
+++ b/security/integrity/ima/ima_template.c
@@ -24,6 +24,7 @@ static struct ima_template_desc builtin_templates[] = {
 	{.name = "ima-sigv2", .fmt = "d-ngv2|n-ng|sig"},
 	{.name = "ima-buf", .fmt = "d-ng|n-ng|buf"},
 	{.name = "ima-modsig", .fmt = "d-ng|n-ng|sig|d-modsig|modsig"},
+	{.name = "ima-ns", .fmt = "userns|d-ng|n-ng|sig"},
 	{.name = "evm-sig",
 	 .fmt = "d-ng|n-ng|evmsig|xattrnames|xattrlengths|xattrvalues|iuid|igid|imode"},
 	{.name = "", .fmt = ""},	/* placeholder for a custom format */
-- 
2.37.3

